File: 0904_FruitIntoBaskets/SlidingWindow.cpp
// Problema: https://imgur.com/a/YkanSAr
// Solución Sliding Window
// Entradas:
    /*
    Caso 1:
    [1,2,1]
    Caso 2: 
    [0,1,2,2]
    Caso 3:
    [1,2,3,2,2]
    */

#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int totalFruit(vector<int>& fruits) {
        int left = 0, best = -1, size = fruits.size();
        // Variables que indican el tipo de fruta que almacena la canasta-
        int bas1 = -1, bas2 = -1;
        for(int right = 0; right < size; right++){
            // El arbol tiene frutas que no se han clasificado en las canastas
            if(fruits[right] != bas1 && fruits[right] != bas2){
                // Se agrega a la canasta (Si se puede)
                if(bas1 == -1) bas1 = fruits[right]; // Canasta 1
                else if(bas2 == -1) bas2 = fruits[right]; // Canasta 2
                else{
                    // NO SE PUEDE AGREGAR (Canastas llenas)
                    // Se quitan las frutas que no formen parte del ultimo arbol revisado
                    bas1 = fruits[right]; // Nuevo arbol
                    bas2 = fruits[right-1]; // Arbol previamente revisado
                    left = right-1; // Se guarda la posicion
                    while(fruits[left-1] == bas2) left--; // Hasta el ultimo arbol del mismo tipo
                }
            }
            // Se valida la mejor distancia (mas arboles)
            best = max(best, right - left + 1);
        }
        // cout << best << endl;
        return best;
    }
};

int main (void){
    Solution casoP;
    vector<int> v1 = {1,2,1}, v2 = {0,1,2,2}, v3 = {1,2,3,2,2};
    cout << casoP.totalFruit(v1) << endl;
    cout << casoP.totalFruit(v2) << endl;
    cout << casoP.totalFruit(v3) << endl;
    return 0;
}

File: 0904_FruitIntoBaskets/SlidingWindow.cs
// Problema: https://imgur.com/a/YkanSAr
// Solución Sliding Window
// Entradas:
    /*
    Caso 1:
    [1,2,1]
    Caso 2: 
    [0,1,2,2]
    Caso 3:
    [1,2,3,2,2]
    */

using System;


var sol = new Solution();
int[] fruits1 = { 1, 2, 1, 2, 3 };
int[] fruits2 = { 0, 1, 2, 2 };
int[] fruits3 = { 1, 2, 3, 2, 2 };
int[] fruits4 = { 3,3,3,1,2,1,1,2,3,3,4 };
Console.WriteLine(sol.TotalFruit(fruits1));
Console.WriteLine(sol.TotalFruit(fruits2));
Console.WriteLine(sol.TotalFruit(fruits3));
Console.WriteLine(sol.TotalFruit(fruits4));
// Console.WriteLine("Hola, mundo!");

public class Solution
{
    public int TotalFruit(int[] fruits)
    {
        int left = 0, best = -1, size = fruits.Length;
        // Variables que usan el tipo de fruta
        int basket1 = -1, basket2 = -1;
        for (int right = 0; right < size; right++)
        {
            if (fruits[right] != basket1 && fruits[right] != basket2)
            {
                if (basket1 == -1) basket1 = fruits[right];
                else if (basket2 == -1) basket2 = fruits[right];
                else
                {
                    basket1 = fruits[right];
                    basket2 = fruits[right - 1];
                    left = right - 1;
                    while (fruits[left - 1] == basket2) left--;
                }
            }
            best = Math.Max(best, right - left + 1);
        }
        // Console.WriteLine();
        return best;
    }
};